### PR TITLE
Add network protocol version to p2p connections

### PR DESCRIPTION
**Pull request by pavel-kokolemin**
I think having a separate field to negotiate the p2p network protocol will make future upgrades easier.
Bitcoin Core uses something like that too: https://github.com/bitcoin/bitcoin/blob/master/src/version.h

### CODE REVIEW

File: p2p/src/net/default_backend/backend.rs [hunk 1]
```diff
@@ -479,17 +483,19 @@ where
...
+                version,
                 receiver_address,
                 handshake_nonce,
-                user_agent,
             } => self.create_peer(
                 peer_id,
                 handshake_nonce,
                 PeerInfo {
                     peer_id,
+                    protocol,
```
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply p2p/src/net/default_backend/backend.rs:498]: Right, the version field is enough. But using a separate filed for the protocol version should be a bit easier to use. There will probably be many node releases, but much fewer protocol upgrades. And this is a very important part, so making it explicit is better.
Also, what if we have third party clients in our p2p network? Why should they have to emulate versions of our node? And which version should they use?
Bitcoin Core also uses two fields for this. The protocol version is sent over the network in the handshake, and the application version is sent in the user agent string. Both values are closely related, but only one is used to decide how to communicate with the peer (the protocol version).

File: p2p/tests/block_announcement.rs [hunk 2]
```diff
@@ -150,15 +151,13 @@ where
     ));
 }
 
-// TODO: Implement announcements resending in partially connected networks.
```
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply p2p/tests/block_announcement.rs:153]: Sure, I just removed the extra TODO copies.

File: p2p/src/net/default_backend/tests.rs [hunk 3]
```diff
@@ -66,7 +66,7 @@ where
     {
         assert_eq!(address, conn2.local_addresses()[0]);
         assert_eq!(peer_info.network, *config.magic_bytes());
-        assert_eq!(peer_info.version, SemVer::new(0, 1, 0));
+        assert_eq!(peer_info.version, *config.version());
```
Comment (TheQuantumPhysicist) [inline p2p/src/net/default_backend/tests.rs:69]: The version should be coming from Cargo... not from config. But maybe this is something for another day.

You can write a TODO in ChainConfig at the version variable stating this.

File: p2p/src/error.rs [hunk 4]
```diff
@@ -22,14 +22,18 @@ use common::{
...
 /// and the peer getting banned.
 #[derive(Error, Debug, PartialEq, Eq)]
 pub enum ProtocolError {
+    #[error("Peer has an unsupported network protocol: {0:?}")]
+    UnsupportedProtocol(NetworkProtocol),
     #[error("Peer is in different network. Our network {0:?}, their network {1:?}")]
     DifferentNetwork([u8; 4], [u8; 4]),
     #[error("Peer has an unsupported version. Our version {0}, their version {1}")]
-    InvalidVersion(SemVer, SemVer),
+    UnsupportedVersion(SemVer, SemVer),
```
Comment (TheQuantumPhysicist) [inline p2p/src/error.rs:36]: Is this really usable?

**Review by TheQuantumPhysicist — APPROVED** [review]

File: p2p/src/error.rs (hunk 4 above)
Comment (pavel-kokolemin) [reply p2p/src/error.rs]: Right, removed, thanks.

File: p2p/src/net/default_backend/tests.rs (hunk 3 above)
Comment (pavel-kokolemin) [reply p2p/src/net/default_backend/tests.rs:69]: I updated `Builder` to use `CARGO_PKG_VERSION`. Is this OK?
```
...
            version: SemVer::try_from(env!("CARGO_PKG_VERSION"))
                .expect("invalid CARGO_PKG_VERSION value"),
```

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**